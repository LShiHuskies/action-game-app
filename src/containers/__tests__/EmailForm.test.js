import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { createMemoryHistory } from 'history';


import {
    Router,
    Switch,
    Route
  } from "react-router-dom";


import rootReducer from '../../reducers';
import EmailForm from '../EmailForm';

const store = createStore(rootReducer, applyMiddleware(thunk));
const memoryHistory = createMemoryHistory();

memoryHistory.entries[0].pathname = '/passwords/forgot';






const renderComponent = () => {
    const helpers = render(
      <Provider store={store}>
        <Router history={memoryHistory}>
          <Switch>
            <Route exact path="/passwords/forgot" component={EmailForm}></Route>
          </Switch>
        </Router>
      </Provider>
    );
    return { ...helpers };
}



describe('<EmailForm />', () => {

  describe('EmailForm Snapshot', () => {
    it('should match snapshot', () => {
        const tree = renderer.create(
            <Provider store={store}>
            <Router history={memoryHistory}>
              <Switch>
                <Route> <EmailForm /> </Route>
              </Switch>
            </Router>
          </Provider>
        );
    
        expect(tree.toJSON()).toMatchSnapshot();
      });
  });

  describe('EmailForm', () => {
    let OriginalConsoleError

    beforeAll(() => {
      OriginalConsoleError = console.error;
      console.error = jest.fn();
    });

    afterAll(() => {
      console.error = OriginalConsoleError;
    });

    describe('Heading Title', () => {
      it('should render a heading title', () => {
        renderComponent();

        expect(screen.getByRole('heading', { level: 5 })).toHaveTextContent('Forgot Your Username or Password?');
      });
    });

    describe('Heading Description', () => {
      it('should render a <p> with the description', () => {
        renderComponent();

        expect(screen.getByText("Provide your email address and we'll send you instructions to reset your account.")).toBeInTheDocument();
      });
    });

    describe('Email', () => {
      describe('Email Input', () => {
        it('should render Two Email Labels', () => {
          renderComponent();
    
          expect(screen.getAllByText('Email')).toHaveLength(2);
        });
      });
    
      describe('Email HelperTexts', () => {
        it('should render email helper text', () => {
          renderComponent();
    
          expect(screen.getByText(/Please enter your email/i)).toHaveClass('MuiFormHelperText-root MuiFormHelperText-contained Mui-required');
        });
      });

      describe('When entering Email in text input', () => {
        it('should allow user to type', () => {
          renderComponent();

          const Email = screen.getByLabelText('Email *');
          const EmailHelperText = screen.getByText('Please enter your email');

          expect(Email).toBeInTheDocument();
          expect(Email.closest("input").value).toEqual("");

          fireEvent.change(Email, { target: { value: "Hello" } });
    
          expect(Email.closest("input").value).toEqual("Hello");
          expect(EmailHelperText).toHaveStyle({ color: 'rgba(0, 0, 0, 0.54)' });
        });

        describe('After entering the Email in Text Input and removing it', () => {
          it('should make the outline of input box and helper text red', () => {
            renderComponent();

            const Email = screen.getByLabelText('Email *');

            expect(Email).toBeInTheDocument();
            expect(Email.closest("input").value).toEqual("");

            fireEvent.change(Email, { target: { value: "Hello" } });
    
            expect(Email.closest("input").value).toEqual("Hello");

            fireEvent.change(Email, { target: { value: "" } });

            expect(Email.closest("input").value).toEqual("");

            const EmailHelperText = screen.getByText('Please enter your email');

            expect(EmailHelperText).toHaveStyle({ color: 'rgb(244, 67, 54)' });
          });
        });
      });
    });

    describe('BACK button', () => {
      it('should render a back button', () => {
        renderComponent();

        expect(screen.getByRole('button', { name: 'Back' })).toBeInTheDocument();
      });

      describe('when clicked', () => {
        it('will go back to the base path', () => {
          renderComponent();

          fireEvent.click(screen.getByRole('button', { name: 'Back' }))

          expect(memoryHistory.location.pathname).toBe('/');
        });
      });
    });
  });
});

