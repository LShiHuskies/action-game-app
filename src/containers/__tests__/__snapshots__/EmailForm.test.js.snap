// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmailForm /> EmailForm Snapshot should match snapshot 1`] = `
Array [
  " ",
  <div
    className="App"
  >
    <header
      className="App-header"
    >
      <h5
        style={
          Object {
            "marginBottom": "0px",
            "marginRight": "150px",
            "marginTop": "0px",
            "width": "100%",
          }
        }
        variant="contained"
      >
         Forgot Your Username or Password? 
      </h5>
      <p
        style={
          Object {
            "fontSize": "13px",
            "margin": "5px",
            "marginLeft": "0px",
          }
        }
      >
        Provide your email address and we'll send you instructions to reset your account.
      </p>
      <div
        className="login"
        style={
          Object {
            "paddingTop": "10px",
          }
        }
      >
        <form
          onSubmit={[Function]}
        >
          <label
            className="form_label"
          >
            Email
          </label>
          <div
            className="MuiFormControl-root MuiTextField-root"
            input={
              Object {
                "name": "email",
                "onBlur": [Function],
                "onChange": [Function],
                "onDragStart": [Function],
                "onDrop": [Function],
                "onFocus": [Function],
                "value": "",
              }
            }
            meta={
              Object {
                "active": false,
                "asyncValidating": false,
                "autofilled": false,
                "dirty": false,
                "dispatch": [Function],
                "error": undefined,
                "form": "EmailForm",
                "initial": undefined,
                "invalid": false,
                "pristine": true,
                "submitFailed": false,
                "submitting": false,
                "touched": false,
                "valid": true,
                "visited": false,
                "warning": undefined,
              }
            }
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
              data-shrink={false}
              htmlFor="standard-error-helper-text"
              id="standard-error-helper-text-label"
            >
              Email
              <span
                aria-hidden={true}
                className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
              >
                 
                *
              </span>
            </label>
            <div
              className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
              onClick={[Function]}
            >
              <input
                aria-describedby="standard-error-helper-text-helper-text"
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiOutlinedInput-input"
                disabled={false}
                id="standard-error-helper-text"
                name="email"
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={true}
                type="text"
              />
              <fieldset
                aria-hidden={true}
                className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
              >
                <legend
                  className="PrivateNotchedOutline-legendLabelled-3"
                >
                  <span>
                    Email
                     *
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              className="MuiFormHelperText-root MuiFormHelperText-contained Mui-required"
              id="standard-error-helper-text-helper-text"
            >
              Please enter your email
            </p>
          </div>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-text"
            disabled={false}
            id="demo"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "height": "40px",
                "margin": "10px",
                "marginTop": "0px",
                "width": "100px",
              }
            }
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              <span
                className="MuiButton-startIcon MuiButton-iconSizeMedium"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
                  />
                </svg>
              </span>
              Back
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
            disabled={false}
            id="submit"
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            style={
              Object {
                "height": "40px",
                "margin": "0px",
                "marginRight": "10px",
                "width": "100px",
              }
            }
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Send
              <span
                className="MuiButton-endIcon MuiButton-iconSizeMedium"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4l-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
                  />
                </svg>
              </span>
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </form>
      </div>
    </header>
  </div>,
  " ",
]
`;
